### PR TITLE
Improve performance of derivative

### PR DESCRIPTION
**Pull request by KirillZubov**
improve performance of derivative

### CODE REVIEW

File: src/pinns_pde_solve.jl [hunk 1]
```diff
@@ -294,25 +294,24 @@ function generate_training_sets(domains,dx,bcs,indvars,depvars,dict_indvars,dict
...
+        push!(train_set, [points...])
     end
 
-    train_bound_set =  Array{Array{Float32,1}}[]
+    train_bound_set = []
     for bt in bound_args
-        _set = Array{Float32}[]
+        _set = []
         span = [get(dict_var_span, b, b) for b in bt]
         for points in Iterators.product(span...)
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl]: don't splat arrays.

File: src/pinns_pde_solve.jl [hunk 2]
```diff
@@ -294,25 +294,24 @@ function generate_training_sets(domains,dx,bcs,indvars,depvars,dict_indvars,dict
...
-            push!(_set, Float32[points...])
+            push!(_set, [points...])
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl]: map instead of push!

File: src/pinns_pde_solve.jl [hunk 3]
```diff
@@ -282,80 +293,49 @@ function generate_training_sets(domains,dx,bcs,indvars,depvars,dict_indvars,dict
     spans = [d.domain.lower:dx:d.domain.upper for (d,dx) in zip(domains,dxs)]
     dict_var_span = Dict([Symbol(d.variables) => d.domain.lower:dx:d.domain.upper for (d,dx) in zip(domains,dxs)])
 
-    train_set = []
-    for points in Iterators.product(spans...)
-        push!(train_set, [points...])
-    end
+    train_set = map(points -> [points...], Iterators.product(spans...))
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:296]: why splat?

File: src/pinns_pde_solve.jl [hunk 4]
```diff
@@ -282,80 +293,49 @@ function generate_training_sets(domains,dx,bcs,indvars,depvars,dict_indvars,dict
...
-            θ = _x[end]
-            εs_dnv = [εs[d] for d in dnv]
-            return _derivative(x,θ,order,εs_dnv,var_num)
-        end
-        _derivative = (x,θ,order,εs_dnv,var_num) ->
-        begin
-            ε = εs_dnv[order]
+            ε = εs[order]
             if order == 1
                 if isuinplace
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:331]: can u ever be in place if we are reverse mode? No I think.

File: src/pinns_pde_solve.jl [hunk 5]
```diff
@@ -282,80 +293,49 @@ function generate_training_sets(domains,dx,bcs,indvars,depvars,dict_indvars,dict
...
-                    return (phi(x+ε,θ) - phi(x-ε,θ))/(2*epsilon)
+                    return (u(x+ε,θ) - u(x-ε,θ))/(2*epsilon)
                 else
-                    return (phi(x+ε,θ)[var_num] - phi(x-ε,θ)[var_num])/(2*epsilon)
+                    return (u(x+ε,θ)[var_num] - u(x-ε,θ)[var_num])/(2*epsilon)
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:334]: you're going to repeat derivative calls. The computation should be re-ordered next PR.

File: src/pinns_pde_solve.jl [hunk 6]
```diff
@@ -282,80 +293,49 @@ function generate_training_sets(domains,dx,bcs,indvars,depvars,dict_indvars,dict
...
-    flat_train_bound_set = collect(Iterators.flatten(train_bound_set))
+    flat_train_bound_set = map(x -> x, Iterators.flatten(train_bound_set))
     train_domain_set =  setdiff(train_set, flat_train_bound_set)
 
     [train_domain_set,train_bound_set,train_set]
 end
 
-function get_phi(chain,isuinplace)
+function get_phi(chain)
+    chain isa Chain && "Chain deprecated, use FastChain"
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:310]: no, we need to support Chain.

File: src/pinns_pde_solve.jl [hunk 7]
```diff
@@ -282,80 +293,49 @@ function generate_training_sets(domains,dx,bcs,indvars,depvars,dict_indvars,dict
...
-            phi = (x,θ) -> first(chain(adapt(typeof(θ),x),θ))
-        else
-            phi = (x,θ) -> chain(adapt(typeof(θ),x),θ)
-        end
+    if isuinplace
+        phi = (x,θ) -> first(chain(adapt(typeof(θ),x),θ))
     else
-        _,re  = Flux.destructure(chain)
-        if isuinplace
-            phi = (x,θ) -> first(re(θ)(adapt(typeof(θ),x)))
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:317]: do you need to destructure here?

File: src/pinns_pde_solve.jl (hunk 6 above)
Comment (KirillZubov) [reply src/pinns_pde_solve.jl:310]: ok

File: src/pinns_pde_solve.jl (hunk 7 above)
Comment (KirillZubov) [reply src/pinns_pde_solve.jl:317]: no

File: src/pinns_pde_solve.jl (hunk 4 above)
Comment (KirillZubov) [reply src/pinns_pde_solve.jl]: yes, "in place" is not needed at all

File: src/pinns_pde_solve.jl (hunk 3 above)
Comment (KirillZubov) [reply src/pinns_pde_solve.jl]: replaced with `collect(points)` instead of splat.

File: src/pinns_pde_solve.jl (hunk 7 above)
Comment (ChrisRackauckas) [reply src/pinns_pde_solve.jl:317]: oh actually I think you do if you're going to use sciml_train over Flux Chain here.

File: src/pinns_pde_solve.jl (hunk 3 above)
Comment (ChrisRackauckas) [reply src/pinns_pde_solve.jl]: what is `points` here: a tuple?

Comment (KirillZubov) [reply src/pinns_pde_solve.jl]: yes

Comment (KirillZubov) [reply src/pinns_pde_solve.jl]: but actually, splat is faster

File: src/pinns_pde_solve.jl (hunk 7 above)
Comment (KirillZubov) [reply src/pinns_pde_solve.jl:317]: I confused with the Flux.destructor from PhysicsInformedNN

Comment (KirillZubov) [reply src/pinns_pde_solve.jl:317]: yeah, it needs here

File: src/pinns_pde_solve.jl [hunk 8]
```diff
@@ -31,11 +31,7 @@ function PhysicsInformedNN(dx,
                            strategy = GridTraining(),
                            kwargs...)
     if init_params === nothing
-        if chain isa FastChain
-            initθ = DiffEqFlux.initial_params(chain)
-        else
-            initθ,re  = Flux.destructure(chain)
-        end
+        initθ = DiffEqFlux.initial_params(chain)
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:34]: is there a test that says this is okay with Flux.Chain? I would think no?

File: src/pinns_pde_solve.jl (hunk 3 above)
Comment (ChrisRackauckas) [reply src/pinns_pde_solve.jl]: Really? Show how: the two should be the same if it's a small tuple. But the bigger thing is, why are tuples generated? It should probably go straight to arrays.

File: src/pinns_pde_solve.jl (hunk 8 above)
Comment (KirillZubov) [reply src/pinns_pde_solve.jl:34]: fastchain = FastChain(FastDense(2,16,Flux.σ),FastDense(16,16,Flux.σ),FastDense(16,1))
chain = Chain(Dense(2,16,Flux.σ),Dense(16,16,Flux.σ),Dense(16,1))

initθ1 = DiffEqFlux.initial_params(chain)
initθ2,_  = Flux.destructure(chain)

initθ1 == initθ2

Comment (KirillZubov) [reply src/pinns_pde_solve.jl:34]: it is the wrapper : `initial_params(f::Chain) = Flux.destructure(f)[1]` 
https://github.com/SciML/DiffEqFlux.jl/blob/b642d5c917b33e2593165a2e1b076f4080eb6a9b/src/fast_layers.jl#L5

File: src/pinns_pde_solve.jl (hunk 3 above)
Comment (KirillZubov) [reply src/pinns_pde_solve.jl]: yes, we should go straight to arrays it would be the best but `Iterators.product(spans...)` create a tuple.

Comment (KirillZubov) [reply src/pinns_pde_solve.jl]:
```julia
    spans = [0.0:0.25:2.0, 0.0:0.25:2.0, 0.0:0.25:2.0]
    julia> @benchmark map(points -> collect(points), Iterators.product(spans...))
    BenchmarkTools.Trial:
      memory estimate:  86.42 KiB
      allocs estimate:  737
      --------------
      minimum time:     51.168 μs (0.00% GC)
      median time:      59.035 μs (0.00% GC)
      mean time:        88.805 μs (21.89% GC)
      maximum time:     68.412 ms (99.81% GC)
      --------------
      samples:          10000
      evals/sample:     1

    julia> @benchmark map(points -> [points...], Iterators.product(spans...))
    BenchmarkTools.Trial:
      memory estimate:  86.42 KiB
      allocs estimate:  737
      --------------
      minimum time:     48.730 μs (0.00% GC)
      median time:      66.870 μs (0.00% GC)
      mean time:        158.667 μs (13.34% GC)
      maximum time:     76.640 ms (99.82% GC)
      --------------
      samples:          10000
      evals/sample:     1
    ```
    uh,  `collect` is a little faster.

**Comment by ChrisRackauckas** [comment]
Is there a test in the test set that solves like Poisson with a Flux Chain? It would be good to test it through the whole system.

**Pull request merged**